src/main/java/org/leleuj/AppHandlerFactory.java

```diff
@@ -28,49 +34,50 @@ public AppHandlerFactory() {
-
-                handler("theForm.html", new FormHandler(formClient));
-                handler("logout.html", new LogoutHandler());
-                handler("index.html", new IndexHandler());
-                handler("callback", new Pac4jCallbackHandler());
-            }
-        });
+    public Handler create(Registry registry) throws Exception {
+        final Saml2Client saml2Client = new Saml2Client();
+        saml2Client.setKeystorePath("resource:samlKeystore.jks");
+        saml2Client.setKeystorePassword("pac4j-demo-passwd");
+        saml2Client.setPrivateKeyPassword("pac4j-demo-passwd");
+        saml2Client.setIdpMetadataPath("resource:testshib-providers.xml");
+
+        final FacebookClient facebookClient = new FacebookClient("145278422258960", "be21409ba8f39b5dae2a7de525484da8");
+        final TwitterClient twitterClient = new TwitterClient("CoxUiYwQOSFDReZYdjigBA",
+                "2kAzunH5Btc4gRSaMr7D7MkyoJ5u1VzbOOzE8rBofs");
+        // HTTP
+        final FormClient formClient = new FormClient(RatpackPac4jDemo.URL + "/theForm.html", new SimpleTestUsernamePasswordAuthenticator());
+        final BasicAuthClient basicAuthClient = new BasicAuthClient(new SimpleTestUsernamePasswordAuthenticator());
+
+        // CAS
+        final CasClient casClient = new CasClient();
+        // casClient.setGateway(true);
+        casClient.setCasLoginUrl("http://localhost:8888/cas/login");
+
+        return chain(registry, (chain) -> chain
+                        .register(r -> {
+                            r.add(ServerErrorHandler.class, new AppServerErrorHandler());
+                            r.add(ClientErrorHandler.class, new AppClientErrorHandler());
+                        })
+
+                        .handler(new Pac4jClientsHandler("callback", formClient, formClient, saml2Client, facebookClient, twitterClient,
+                                basicAuthClient, casClient))
+                        .handler("", new DefaultRedirectHandler())
+
+                        .prefix("facebook", new AuthenticatedPageChain("FacebookClient"))
+                        .prefix("twitter", new AuthenticatedPageChain("TwitterClient"))
+                        .prefix("form", new AuthenticatedPageChain("FormClient"))
+                        .prefix("basicauth", new AuthenticatedPageChain("BasicAuthClient"))
+                        .prefix("cas", new AuthenticatedPageChain("CasClient"))
+                        .prefix("saml2", new AuthenticatedPageChain("Saml2Client"))
+
+                        .handler("theForm.html", new FormHandler(formClient))
+                        .handler("logout.html", new LogoutHandler())
+                        .handler("index.html", new IndexHandler())
+                        .handler("callback", new Pac4jCallbackHandlerBuilder().build())
+        );
     }
 
-    private class AuthenticatedPageChain extends ChainAction {
+    private class AuthenticatedPageChain implements Action<Chain> {
```

Comment:
`*-pac4j` libraries give us a great overview between all frameworks: Play vs Ratpack vs Undertow vs Spark for example...